### DISCUSSION TITLE
Multithreaded output redirection

### DISCUSSION THREAD

**Discussion by joeletho** [JustWhit3/osmanip#21] [Ideas]
Hi @JustWhit3,

I'd like your input on my current dilemma with redirecting output on multiple threads. Here's the gist: each thread swaps the std::cout buffer for their own buffer when begin() is called and written to the file when end() is called. This works as expected on a single thread, but gets out of whack when multiple threads call begin() and end() because they're all fighting for the std::cout buffer and swapping things around. 

Here's some ideas I have that may resolve the issue:
- Override cout inside the std namespace to control the outflow of information (i.e. if begin() has been called then update that buffer, otherwise send it to cout like normal.) I'm not stoked on using the std namespace, but it could be implemented to not cause any issues when used normally. 
- Use an observer (like OutputRedirector to control the flow. However, this would require using something other than std::cout within the osm namespace, at least. Say, osm::cout. This means the code base would need to be updated once the class has been implemented, but makes redirecting output trivial. 
- The final idea, which is also my least favorite, is making threads wait until the first thread in the queue has called end(), which would then notify the next thread to begin. Not ideal and essentially defeats the purpose of multithreaded work. 

I haven't tried any of this yet, so at this point it's all theoretical, but should still work.

Let me know your thoughts.

**Comment by JustWhit3** [comment]
Hi @joeletho your ideas seem interesting. Looking at your proposals one-by-one:

1) This is an interesting idea, but the problem is that overriding `std::cout` may bring some portability problems and I would prefer to avoid this option. I am of the "school" that std stuff has to remain as it is.
2) Why this option would require something other than `std::cout`? However doing a proper `osm::cout` may not be a bad idea. In any case I would better understand what are the impacts of this choice, how many pieces of codes do we need to change to implement this? At the moment it seems the more feaseble option.
3) I would avoid this idea since, as you said, it would defeats the purpose of multithreading work.

Cheers,
Gianluca.

**Comment by joeletho** [comment]
Just thinking out loud (because I'm away from my computer), I think adding osm::cout to the library would add even more functionality to the library. That way, std::cout would stay completely independent of osm::cout, and osm::cout would use a different buffer for each thread it's called from. So, we retain the static functionality of std::cout, but have the ability to do all sorts of things with osm::cout, I.e. osm::cout.flush, osm::cout.redirect, etc..

**Reply by joeletho** [reply]
Hi @JustWhit3, thank you for the reply.

I agree that options 1 and 3 should be disregarded.

To address your questions:
> Why this option would require something other than ```std::cout```? 
>
This is because ```std::cout``` can only use one buffer at a time. I'll try to illustrate the issue with pseudo code:
```
thread1:
    // swaps buffers with std::cout to collect all calls to std::cout 
    // and stores a pointer to the original std::cout buffer
    redirect.begin() 

    // does some stuff

thread2:
    // swaps buffers with thread1 to collect all calls to std::cout 
    // and stores a pointer to thread1's buffer
    redirect.begin()

thread1:
    // ...
    redirect.end()
    // restores std::cout buffer with the original std::cout buffer
    // and writes nothing to the file because thread2 replaced
    // the buffer

thread2:
    // ...
    redirect.end()
    // restores std::cout with thread1's buffer and writes nothing
    // to the file because thread1 replaced the buffer with the
    // original buffer
```
This is what I've experienced in my tests so far. Because of this, it would be advantageous to add ```osm::cout``` because ```osm::cout``` the backend can get the thread id and write to the appropriate buffer automatically.  At least, that's the plan. I haven't yet implemented it yet, so I could be wrong.

> how many pieces of codes do we need to change to implement this?
>
If this works, then replacing all instances of ```std::cout``` with ```osm::cout``` and a new header include like ```<osmanip/iostream.hpp>```, ```<osmanip/osmio.hpp>``` or whatever is to your liking. It _should_ be possible to get it to work just like ```std::cout``` which states 
>These objects are guaranteed to be initialized during or before the first time an object of type [std::ios_base::Init](https://en.cppreference.com/w/cpp/io/ios_base/Init) is constructed...
>
So, if we can replicate the ```std::cout``` constructor, ```osm::cout``` will be initialized the first time ```osm::cout``` is called. 

I'll see what I can come up with and get a working prototype within the next few days or so. This is approach is more invasive than I would prefer, but I'm just not sure how else to go about it.

**Reply by JustWhit3** [reply]
Hi @joeletho ,

Your idea seems interesting and maybe it is the most suitable way to solve this problem. I agree with all the things you said and if you can start preparing a prototype or something similar to test it would be a good idea.

Just a few reminders during your development:

- Maybe it would be more appropriate to store the new header into `<osmanip/utility/iostream.hpp>` without the needing of creating a new subdir for this. What do you think?
- Please, be sure that any functionality or performance feature would be compromised with this new addition. Portability and performances are fundamental for this library.

**Reply by joeletho** [reply]
Hi @JustWhit3,

I did as you suggested and put the header into ```<osmanip/utility/iostream.hpp>```. I feel like it's a nice fit. And I will continue to keep the functionality and performance of the feature in mind. I don't want to overcomplicate things and end up breaking something.

Of course implementing this feature is more complicated than I had hoped, and has turned out to be a quite a beast. While I wanted to have a working prototype in a few days, it's going to take a bit longer to get right.  I'll provide another update on this once I have something to share.

Best,
Joel

**Reply by JustWhit3** [reply]
Perfect! I will wait for your updates.

Cheers,
Gianluca.
